### PR TITLE
Protobuf round trip unit test

### PR DESCRIPTION
**Pull request by nking-1**
Adds a unit test that does a round-trip serialization then deserialization of a grammar. This way we can cut out the server entirely, hopefully resulting in a more reliable test.

### CODE REVIEW

File: tests/unit/test_protobuf.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    Byte,
+    ByteRange,
+    GrammarFunction,
+    Join,
+    ModelVariable,
+    Select,
+)
+
+
+def compare_grammars(g1: GrammarFunction, g2: GrammarFunction) -> bool:
```
Comment (hudson-ai) [inline tests/unit/test_protobuf.py:22]: Feels more or less worthy of being the `__eq__` on grammars!

Rather than hard-coding the fields of each of these classes, could we reach into their `__slots__` in order to make this a bit more robust to changes their implementations? Without actually implementing this as `__eq__`, you'll still have to special-case `values` though.

Just a thought :)

Comment (nking-1) [reply tests/unit/test_protobuf.py:23]: I like the idea of using slots instead of hard coding the names of the properties. If we move the code for `compare_grammars` to implement `__eq__`, would we be implementing `__eq__` for the `GrammarFunction` base class, or somewhere else?

Comment (hudson-ai) [reply tests/unit/test_protobuf.py:23]: I think it could go as low level as the `Function` base class, but I'm actually a bit hesitant to implement `__eq__`. Because there is a contract that `a == b` implies `hash(a) == hash(b)`, providing a non-default impl of eq makes objects non-hashable until you provide an impl of that too. I'm fairly certain we use their hashability, so this could potentially break a lot of things if we're not careful

**Review by riedgar-ms — COMMENTED** [review]
Thanks for working on this!

**Comment by Harsha-Nori** [comment]
I like the idea of testing the protobuf explicitly, and having a grammar equality operator is very useful! I think @mmoskal wants to eventually move to a purely JSON based serialization format, so would just be curious about his thoughts there before we double down too hard on the protobuf infrastructure

**Comment by hudson-ai** [comment]
@Harsha-Nori even if we are planning on switching over to JSON, I don't think that adding a test for round-trip serialization/deserialization is walking too far down the protobuf rabbit hole. At the bare minimum, inclusion of this test will force us to make a conscious decision about these same behaviors in JSON land.

TLDR: approved from me but I'll wait on some more voices before merging :)

**Review by hudson-ai — APPROVED** [review]

**Comment by Harsha-Nori** [comment]
Sounds good to me! Thanks @nking-1 :)
